addons/udes_security/controllers/main.py

```diff
@@ -5,12 +5,100 @@
+        so instead do a direct query to get the relevant id.
+
+        All models on the `action` Reference field on `ir.ui.menu` have the same
+        `_sequence` parameter, so there is no risk of colliding ids from separate models.
+
+        They are stored in the db like "ir.actions.act_window,90" or "ir.actions.server,242"
+        due to being a Reference field, so perform a fuzzy search on %,<id>% to get it
+
+        Be extra safe and ensure `action_id` is an int, as if the user enters text in the URL
+        for action= param, it will come through to here as a string
+
+        :param: action_id: int() for an action (ir.actions.server, ir.actions.act_window, etc..)
+
+        :return: ir.ui.menu(x,) recordset, will be empty if a non existent action_id is passed in
+        """
+        IrUiMenu = request.env["ir.ui.menu"]
+
+        if not isinstance(action_id, int):
+            raise ValidationError("Invalid action ID specified.")
+        # Escape the leading and trailing static % symbols with %,
+        # as we don't want to string format them in the execute() call
+        query = "SELECT id FROM ir_ui_menu WHERE action LIKE '%%,%s%%'"
+        request.env.cr.execute(query, (action_id,))
+        result = request.env.cr.fetchone()
+        return IrUiMenu.sudo().browse(result)
+
+    def _user_has_access_right(self, action, menu, user):
+        """
+        Check user access rights against action, menu, and menu parent.
+
+        Start at the provided action & menu, and recursively call, passing in the menu parent until either:
+        1.) The user has no groups which are specified against the action or menu (return False)
+        2.) The user has groups specified against the action or menu, and no parent exists (return True)
+          2.2) There are no groups specified against the action or menu, and no parent exists (return True)
+
+        This ensures that the user has access at least every menuitem in the chain of menuitems
+        leading to the menu/action they are trying to access, following the same logic of the web UI.
+
+        :param: action: ir.actions.act_window(x,) recordset
+        :param: menu: ir.ui.menu(x,) recordset
+        :param: user: res.users(x,) recordset
+
+        :return: bool() whether or not the user should be granted access
+        """
+        # TODO(palabaster): This line makes up 25% of this entire functions cpu time, any way we can improve?
+        parent = menu.parent_id
+        group_ids_for_access = []
+        # ir.actions.client, ir.actions.act_url don't have groups_id parameter
+        if hasattr(action, "groups_id"):
+            group_ids_for_access += action.groups_id.ids
+        # Can have actions defined in the system with no menu associated. These should be blocked.
+        if not menu:
+            return False
+        group_ids_for_access += menu.groups_id.ids
+        user_has_any_groups = any(g_id in user.groups_id.ids for g_id in group_ids_for_access)
```

Comment:
will check

Reply:
`user.groups_id.ids` runs this query
```sql
SELECT
    res_groups_users_rel.uid,
    res_groups_users_rel.gid
FROM
    res_groups_users_rel,
    "res_groups"
    LEFT JOIN "ir_translation" AS "res_groups__name" ON (
        "res_groups"."id" = "res_groups__name"."res_id"
        AND "res_groups__name"."type" = 'model'
        AND "res_groups__name"."name" = 'res.groups,name'
        AND "res_groups__name"."lang" = 'en_GB'
        AND "res_groups__name"."value" != ''
    )
WHERE
    1 = 1
    AND res_groups_users_rel.uid IN (2)
    AND res_groups_users_rel.gid = res_groups.id
ORDER BY
    COALESCE("res_groups__name"."value", "res_groups"."name") OFFSET 0
```

Reply:
i'll push it up to parent function too so we don't call it every time round

Reply:
didn't account for postgres caching..

re-ran on a clean session and was
 Planning Time: 0.201 ms
 Execution Time: 0.075 ms

still faster though, and now we aren't running it every parent menuitem

Reply:
`user.groups_id.ids` runs this query
```sql
SELECT
    res_groups_users_rel.uid,
    res_groups_users_rel.gid
FROM
    res_groups_users_rel,
    "res_groups"
    LEFT JOIN "ir_translation" AS "res_groups__name" ON (
        "res_groups"."id" = "res_groups__name"."res_id"
        AND "res_groups__name"."type" = 'model'
        AND "res_groups__name"."name" = 'res.groups,name'
        AND "res_groups__name"."lang" = 'en_GB'
        AND "res_groups__name"."value" != ''
    )
WHERE
    1 = 1
    AND res_groups_users_rel.uid IN (2)
    AND res_groups_users_rel.gid = res_groups.id
ORDER BY
    COALESCE("res_groups__name"."value", "res_groups"."name") OFFSET 0
```
Planning Time: 1.149 ms
Execution Time: 0.276 ms

on a clean postgres session (to avoid caching), can do this which gives same result:
```sql
SELECT ARRAY_AGG(gid) FROM res_groups_users_rel WHERE res_groups_users_rel.uid in (2)
```
Planning Time: 0.201 ms
Execution Time: 0.075 ms

Changed to run once outside the recursive function